service/matching/tasklist/adaptive_scaler.go

```diff
@@ -0,0 +1,255 @@
+	if !atomic.CompareAndSwapInt32(&a.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+	a.cancel()
+	a.wg.Wait()
+}
+
+func (a *adaptiveScalerImpl) runPeriodicLoop() {
+	defer a.wg.Done()
+	timer := a.timeSource.NewTimer(a.config.AdaptiveScalerUpdateInterval())
+	defer timer.Stop()
+	for {
+		select {
+		case <-a.ctx.Done():
+			return
+		case <-timer.Chan():
+			a.logger.Debug("timer")
+			a.run()
+			timer.Reset(a.config.AdaptiveScalerUpdateInterval())
+		}
+	}
+}
+
+func (a *adaptiveScalerImpl) run() {
+	if !a.config.EnableAdaptiveScaler() || !a.config.EnableGetNumberOfPartitionsFromCache() {
+		return
+	}
+	qps := a.qpsTracker.QPS()
+	partitionConfig := a.getPartitionConfig()
+	// adjust the number of write partitions based on qps
+	numWritePartitions := a.adjustWritePartitions(qps, partitionConfig.NumWritePartitions)
+	// adjust the number of read partitions
+	numReadPartitions := a.adjustReadPartitions(partitionConfig.NumReadPartitions, numWritePartitions)
+
+	if numReadPartitions == partitionConfig.NumReadPartitions && numWritePartitions == partitionConfig.NumWritePartitions {
+		return
+	}
+	a.logger.Info("update the number of partitions", tag.CurrentQPS(qps), tag.NumReadPartitions(numReadPartitions), tag.NumWritePartitions(numWritePartitions))
+	a.scope.IncCounter(metrics.CadenceRequests)
+	err := a.tlMgr.UpdateTaskListPartitionConfig(a.ctx, &types.TaskListPartitionConfig{
+		NumReadPartitions:  numReadPartitions,
+		NumWritePartitions: numWritePartitions,
+	})
+	if err != nil {
+		a.logger.Error("failed to update task list partition config", tag.Error(err))
+		a.scope.IncCounter(metrics.CadenceFailures)
+	}
+}
+
+func (a *adaptiveScalerImpl) getPartitionConfig() *types.TaskListPartitionConfig {
+	partitionConfig := a.tlMgr.TaskListPartitionConfig()
+	if partitionConfig == nil {
+		partitionConfig = &types.TaskListPartitionConfig{
+			NumReadPartitions:  1,
+			NumWritePartitions: 1,
+		}
+	}
+	return partitionConfig
+}
+
+func (a *adaptiveScalerImpl) adjustWritePartitions(qps float64, numWritePartitions int32) int32 {
+	upscaleThreshold := float64(a.config.PartitionUpscaleRPS())
+	downscaleFactor := a.config.PartitionDownscaleFactor()
+	downscaleThreshold := float64(numWritePartitions-1) * upscaleThreshold * downscaleFactor / float64(numWritePartitions)
+	a.scope.UpdateGauge(metrics.EstimatedAddTaskQPSGauge, qps)
+	a.scope.UpdateGauge(metrics.TaskListPartitionUpscaleThresholdGauge, upscaleThreshold)
+	a.scope.UpdateGauge(metrics.TaskListPartitionDownscaleThresholdGauge, downscaleThreshold)
+
+	result := numWritePartitions
+	if qps > upscaleThreshold {
+		if !a.overLoad {
+			a.overLoad = true
+			a.overLoadStartTime = a.timeSource.Now()
+		} else if a.timeSource.Now().Sub(a.overLoadStartTime) > a.config.PartitionUpscaleSustainedDuration() {
+			result = getNumberOfPartitions(numWritePartitions, qps, upscaleThreshold)
+			a.overLoad = false
+		}
+	} else {
+		a.overLoad = false
```

Comment:
we can discuss this offline, but I think with `downscaleFactor`, we can handle fluctuation. The default factor is 0.75, which means the number of partitions won't change unless the traffic drop by 25%.

Reply:
I think we need matching simulator to validate different options of scale formula. I assume you will update simulation next and iterate on this. If so this looks like a good start. Can you confirm?

Reply:
It doesn't fit the simulation framework because the output of simulation tests assume that the number of partitions don't change. I can run bench tests instead.

Reply:
Simulation framework can be enhanced to support this. Feedback loop is faster in simulations and also more repeatable than benchmarking in dev env so let's invest in this.